cpp/density/RDF.cc

```diff
@@ -28,262 +27,187 @@ RDF::RDF(float rmax, float dr, float rmin)
-    memset((void*)m_bin_counts.get(), 0, sizeof(unsigned int)*m_nbins);
-    m_avg_counts = std::shared_ptr<float>(new float[m_nbins], std::default_delete<float[]>());
-    memset((void*)m_avg_counts.get(), 0, sizeof(float)*m_nbins);
-    m_N_r_array = std::shared_ptr<float>(new float[m_nbins], std::default_delete<float[]>());
-    memset((void*)m_N_r_array.get(), 0, sizeof(unsigned int)*m_nbins);
+    m_pcf_array = util::returnEmptyArray<float>(m_nbins);
+    m_bin_counts = util::returnEmptyArray<unsigned int>(m_nbins);
+    m_avg_counts = util::returnEmptyArray<float>(m_nbins);
+    m_N_r_array = util::returnEmptyArray<float>(m_nbins);
 
     // precompute the bin center positions and cell volumes
     m_r_array = std::shared_ptr<float>(new float[m_nbins], std::default_delete<float[]>());
-    m_vol_array = std::shared_ptr<float>(new float[m_nbins], std::default_delete<float[]>());
-    memset((void*)m_vol_array.get(), 0, sizeof(float)*m_nbins);
-    m_vol_array2D = std::shared_ptr<float>(new float[m_nbins], std::default_delete<float[]>());
-    memset((void*)m_vol_array2D.get(), 0, sizeof(float)*m_nbins);
-    m_vol_array3D = std::shared_ptr<float>(new float[m_nbins], std::default_delete<float[]>());
-    memset((void*)m_vol_array3D.get(), 0, sizeof(float)*m_nbins);
+    m_vol_array = util::returnEmptyArray<float>(m_nbins);
+    m_vol_array2D = util::returnEmptyArray<float>(m_nbins);
+    m_vol_array3D = util::returnEmptyArray<float>(m_nbins);
+
     for (unsigned int i = 0; i < m_nbins; i++)
-        {
+    {
         float r = float(i) * m_dr + m_rmin;
-        float nextr = float(i+1) * m_dr + m_rmin;
-        m_r_array.get()[i] = 2.0f / 3.0f * (nextr*nextr*nextr - r*r*r) / (nextr*nextr - r*r);
-        m_vol_array2D.get()[i] = M_PI * (nextr*nextr - r*r);
-        m_vol_array3D.get()[i] = 4.0f / 3.0f * M_PI * (nextr*nextr*nextr - r*r*r);
-        }
+        float nextr = float(i + 1) * m_dr + m_rmin;
+        m_r_array.get()[i] = 2.0f / 3.0f * (nextr * nextr * nextr - r * r * r) / (nextr * nextr - r * r);
+        m_vol_array2D.get()[i] = M_PI * (nextr * nextr - r * r);
+        m_vol_array3D.get()[i] = 4.0f / 3.0f * M_PI * (nextr * nextr * nextr - r * r * r);
+    }
 
-    }  // end RDF::RDF
+} // end RDF::RDF
 
 RDF::~RDF()
+{
+    for (tbb::enumerable_thread_specific<unsigned int*>::iterator i = m_local_bin_counts.begin();
+         i != m_local_bin_counts.end(); ++i)
     {
-    for (tbb::enumerable_thread_specific<unsigned int *>::iterator i = m_local_bin_counts.begin(); i != m_local_bin_counts.end(); ++i)
-        {
-        delete[] (*i);
-        }
+        delete[](*i);
     }
+}
 
 //! \internal
 //! CumulativeCount class to perform a parallel reduce to get the cumulative count for each histogram bin
 class CumulativeCount
+{
+private:
+    float m_sum;
+    float* m_N_r_array;
+    float* m_avg_counts;
+
+public:
+    CumulativeCount(float* N_r_array, float* avg_counts)
```

Comment:
This entire class is a holdover from TBB before lambdas. This should definitely be rewritten.
https://www.threadingbuildingblocks.org/docs/help/reference/algorithms/parallel_scan_func.html

Reply:
For some reason, it seems that tbb on Ubuntu cannot find the template for
```
template<typename Range, typename Value, typename Scan, typename Combine>
Value parallel_scan( const Range& range, const Value& identity,
                     const Scan& scan, const Combine& combine
                     [, partitioner] );
```
The implementation using this version fails to compile on Ubuntu containers on circleci, but compiles on my computer (Mac) and arch container.

Reply:
The reason is that this `parallel_scan` was added for the 2018 update 1 for `tbb` but the default on Ubuntu is using the 2017 version.